### DISCUSSION TITLE
Langchain integration

### DISCUSSION THREAD

**Discussion by batrasrishti** [explosion/spacy-llm#281] [Q&A] [answered]
Hello,
1) Can we use ggml models with spacy_llm using Llamacpp from langchain, if yes then can you please guide how that can be done.
2) Please can you tell me what are the things with '@' prefix called, for example @llm_models, @llm_queries?
3) In the documentation it is written, @llm_models = "langchain.OpenAI.v1"  I don't know what I should be putting at the right side for a model on HuggingFaceHub.

**Comment by rmitsch** [comment, marked as answer]
Hi @batrasrishti!

1. You can pass on all `langchain` config parameter to the `langchain` model via the `spacy-llm` config. I've never tried .ggml models via `langchain`, but I don't see a reason why this shouldn't work. If you encounter a particular problem, let us know.
2. The `@` is used for _registries_. See the [docs on the spaCy config system](https://spacy.io/usage/training#config).
3. Any of the handles for the supported HuggingFace models, e. g. `spacy.Lllama2.v1`. See the [docs](https://spacy.io/api/large-language-models#models).

**Reply by batrasrishti** [reply]
1) Suppose I want to use a model like this, llm = Llama(model_path="llama-2-7b-chat.Q4_K_S.gguf", n_ctx=2048, n_batch=512)
I am confused about how to put in these type of settings for the LLM in the config file?
2) Understood, thank you.
3) Point 3 was for HuggingFace models for example: google/flan-t5-xxl which are still not supported by spacy.

**Reply by rmitsch** [reply]
1. You pass these arguments in [your `config` attribute in your `spacy-llm` config](https://spacy.io/api/large-language-models#langchain-models). All these arguments are passed on to `langchain`.
2. :+1: 
3. If you want us to support a particular HF model, feel free to open an issue with a feature request! We'll look into it :slightly_smiling_face:

**Reply by pacoBOeT** [reply]
Hi, it would be fantastic to have Falcon 40b (not the instruct version) as a supported HF model in spacy_llm. Could it be possible? Is there any other way to implement in spacy a non supported HF model with langchain? I may need to use other models from HF and wouldn't want to be annoying you each time. Thank you so much!

**Reply by rmitsch** [reply]
You can always use the `langchain` integration to run any model supported by `langchain` - if that's the case for Falcon 40b, you can use it in `spacy-llm`.

**Comment by OzielLeos** [comment]
Hello batrasrishti,

Yes, you can use models with spacy_llm by leveraging Llamacpp from Langchain! Below, I'll show you how to set this up and provide an example of using it for Named Entity Recognition (NER).

Here’s the langchain_llm_ner.cfg configuration file that outlines how to load the model and set up NER:
# langchain_llm_ner.cfg
```ini
[nlp]
lang = "en"
pipeline = ["llm"]
[components]
[components.llm]
factory = "llm"
[components.llm.task]
@llm_tasks = "spacy.NER.v3"
labels = ["PERSON", "ORGANISATION", "LOCATION"] 
[components.llm.model]
@llm_models = "langchain.LlamaCpp.v1"
name =  "mistralai_mistral-7b-instruct-v0.1"
config = {
    "model_path": "models/mistral-7b-instruct-v0.1.Q4_0.gguf",
    "temperature": 0.0 
    }
```
# Python code
```python
from spacy_llm.util import assemble
nlp = assemble("langchain_llm_ner.cfg")
content = "Jack and Jill rode up the hill in Les Deux Alpes"
doc = nlp(content)
print([(ent.text, ent.label_) for ent in doc.ents])
```

**Reply by ianbstewart** [reply]
Hello! When I try to load a `llama-cpp` model using this config, I get an registry error (see below). It looks like `llama-cpp` isn't recognized internally by `spacy-llm`. Is this a version problem? I'm using spacy version 3.5.3.

```
RegistryError: [E893] Could not find function 'langchain.LlamaCpp.v1' in function registry 'llm_models'. If you're using a custom function, make sure the code is available. If the function is provided by a third-party package, e.g. spacy-transformers, make sure the package is installed in your environment.

Available names: spacy.Ada.v1, spacy.Ada.v2, spacy.Azure.v1, spacy.Azure.v2, spacy.Babbage.v1, spacy.Babbage.v2, spacy.Claude-1-0.v1, spacy.Claude-1-0.v2, spacy.Claude-1-2.v1, spacy.Claude-1-2.v2, spacy.Claude-1-3.v1, spacy.Claude-1-3.v2, spacy.Claude-1.v1, spacy.Claude-1.v2, spacy.Claude-2.v1, spacy.Claude-2.v2, spacy.Claude-instant-1-1.v1, spacy.Claude-instant-1-1.v2, spacy.Claude-instant-1.v1, spacy.Claude-instant-1.v2, spacy.Code-Davinci.v1, spacy.Code-Davinci.v2, spacy.Command.v1, spacy.Command.v2, spacy.Curie.v1, spacy.Curie.v2, spacy.Davinci.v1, spacy.Davinci.v2, spacy.Dolly.v1, spacy.Falcon.v1, spacy.GPT-3-5.v1, spacy.GPT-3-5.v2, spacy.GPT-3-5.v3, spacy.GPT-4.v1, spacy.GPT-4.v2, spacy.GPT-4.v3, spacy.Llama2.v1, spacy.Mistral.v1, spacy.NoOp.v1, spacy.OpenLLaMA.v1, spacy.PaLM.v1, spacy.PaLM.v2, spacy.StableLM.v1, spacy.Text-Ada.v1, spacy.Text-Ada.v2, spacy.Text-Babbage.v1, spacy.Text-Babbage.v2, spacy.Text-Curie.v1, spacy.Text-Curie.v2, spacy.Text-Davinci.v1, spacy.Text-Davinci.v2, spacy.Text-Davinci.v3
```
